README.md

```diff
@@ -91,6 +91,18 @@ class MyComponent extends Component {
 }
 ```
 
+# Code organization
+
+## Ducks
```

Comment:
As explained first by [erikras](https://github.com/erikras/ducks-modular-redux), ducks (like re-dux) is the name he gave to speak about modules in redux. [necolas](http://nicolasgallagher.com/redux-modules-and-code-splitting/) (ex twitter lite) speaks again about redux modules that have been implemented at twitter.

They both call this code collocation redux a module instead of _ducks_ as this name is self-explained.

Redux modules then, do not have anything to deal with React/Preact or any libraries components' components.

That is true that components are called dumb or smart, component or container, component or connected component in association with a flux architecture such as redux. But for testing and for code organizing, I guess they worth to be separated:

```
src
├── components
│   └── NotificationView
└── data
    └── notifications
```

I know it is not the pattern used in others Cozy applications, but I would like to take the chance to give this opinion:

- modules should be called "modules" in documentation and not _ducks_
- components should be separated from data logic, like redux modules.

Reply:
I like better the name `modules`.

> components should be separated from data logic, like redux modules.


This to me is explained here : 

> Dumb components should be outside ducks (as they are not concerned with the store and could be extracted to the UI library) but their connected couterpart should be exported from the index.js of the duck.

Do you have something more specific in mind ? Would you have the connected counterpart outside the module ?

Reply:
You speak about _dumb components_, like if _smart or connected components_ could be in a redux modules. My point was not to collocated any components logic with any redux/data logic. Connected components are just components with injected `props`. Props could be injected by anything included a redux state but not only.
Not collocating components and redux let the flux architecture be more explicit: components send actions to update stores, stores give state slices to components. Then because everything stays far away from each other, we could guess developers will not couple things too much. Then they will be able to write data unit tests to cover data API and UI unit tests to cover components trees. (cf. https://hacks.mozilla.org/2018/04/testing-strategies-for-react-and-redux/)

Reply:
Totally in favor of the name _modules_, as I think that the word _ducks_ was a confusing concept which motivated @sebn [on Mattermost](https://mattermost.cozycloud.cc/test-team/pl/yw91cf61apg6tkpefgpa47sh6r) question this morning.

I have no opinion about components location, but I feel that putting them outside our redux modules is a cleaner solution. In Store or Collect, we have components almost everywhere in the app and it sometime a bit confusing to find them.

Reply:
> You speak about dumb components, like if smart or connected components could be in a redux modules.

Reply:
I like the `connect` export from `data/greetings/index.js`. It could fix a not-so-good pattern in Collect, as both the default browser view and the intent view need to be connected to the redux store, in the exact same way.

Reply:
While I'm totally in favor of the name `modules` (in part because I think redux will be less often needed with the new context API, and thus `ducks` is less desirable), I disagree with your idea of putting components related to the module functionality outside of the module. I get that it wasn't what erikras or necolas had in mind when talking about ducks or redux modules, but I still think it's better to collocate related components with the redux (or context) actions & reducers in the module : if you don't, then we're back to organising the code into layers instead of features, and I think it makes the apps architecture more difficult to understand. 
(Note that redux logic and components must still be separed in the module, like:
```txt
src
├── notifications
│   └── components
│        └── __tests__
│        └── NotificationView
│    └── redux 
```)
This organization doesn't hinder testing, you can still test your redux parts and your dumb components. The coupling is minimal because it's only the code that connects the components in the `index.jsx`, and I prefer to isolate that kind of plumbing code that hinders readability. And the flux architecture __of the module__ is more explicit because it's not mixed with the rest of the slices.
Of course, the whole app can't be written like this, we'll still need some kind of "main" module that assemble connected components into complete views, and there are probably some edge cases. But I think it's better to not slice our code into logical layers when possible.

Reply:
Whatever the superior organization which can be functionalities or localized parts of the application (header, footer, navbar, etc.) or even mini applications (within a dashboard for example), you propose a decoupling between the data and the components and this is entirely suitable for the proposals detailed above (testability, partitioning, weak coupling).

It remains the case where you share code between several parts of your application without being able to export a version to a component library (like a contact list very specific to an specialized application) and which feeds on different data depending on where in the application the component is used (my global list, my group, intent, etc.) But this may not need to be specified in a guideline.

Reply:
Ok so I feel like we have a consensus over there. So our guideline should look to something like the following, with the example of Greeting above:

```
src
├── greetings
│   └── components
│        └── __tests__
│        └── Greeting.jsx
│    └── redux
│        └── index.js
```
If we all agree I suggest @ptbrowne to upgrade the current PR and merge all of this.